src/lightmap.cpp

```diff
@@ -186,6 +186,86 @@ void map::apply_character_light( player &p )
     }
 }
 
+// This function raytraces starting at the upper limit of the simulated area descending
+// toward the lower limit. Since it's sunlight, the rays are parallel.
+// Each layer consults the next layer up to determine the intensity of the light that reaches it.
+// Once this is complete, additional operations add more dynamic lighting.
+void map::build_sunlight_cache( int zlev )
+{
+    level_cache &map_cache = get_cache( zlev );
+    auto &lm = map_cache.lm;
+    // Grab illumination at ground level.
+    const float outside_light_level = g->natural_light_level( 0 );
+    const float inside_light_level = ( outside_light_level > LIGHT_SOURCE_BRIGHT ) ?
+                                     LIGHT_AMBIENT_LOW + 1.0 : LIGHT_AMBIENT_MINIMAL;
+    // Handling when z-levels are disabled is based on whether a tile is considered "outside".
+    if( !zlevels ) {
+        const auto &outside_cache = map_cache.outside_cache;
+        for( int x = 0; x < MAPSIZE_X; x++ ) {
+            for( int y = 0; y < MAPSIZE_Y; y++ ) {
+                if( outside_cache[x][y] ) {
+                    lm[x][y].fill( outside_light_level );
+                } else {
+                    lm[x][y].fill( inside_light_level );
+                }
+            }
+        }
+        return;
+    }
+    // If uppermost level, just apply weather illumination since there's no opportunity
+    // for light to be blocked.
+    if( zlev == OVERMAP_HEIGHT ) {
+        for( int x = 0; x < MAPSIZE_X; x++ ) {
+            for( int y = 0; y < MAPSIZE_Y; y++ ) {
+                lm[x][y].fill( outside_light_level );
+            }
+        }
+        return;
+    }
+
+    // Replace this with a calculated shift based on time of day and date.
+    // At first compress the angle such that it takes no more than one tile of shift per level.
```

Comment:
You're saying it must be at most one file of shift, but the offset is an integer, so that means the only options are -1, 0, and 1 in each coordinate.  Do you intend to support narrower angles than that?  It will make things more complicated.

Reply:
That is what it means, and yes it will be a little more complicated.

I have some functions @KorGgenT tracked down for me to calculate solar azimuth and elevation, which I'll need to compress to I think a 9deg arc to meet this prerequisite.
This angle is used to calculate a vector (math vector not data structure vector) indicating the direction and magnitude of the shadow.
Then this vector is used to calculate a line descending from max height to max depth (bresenham-wise). We compare the offset for a given layer to the one above it, and if there is a difference, that layer is shifted one tile relative to it's antecedent.

I'll have to evaluate if this is a meaningful thing to do (if you almost never see shadows, there's no point), if my plan is workable, and if it is likely to cause artifacts.

There's a whole other thing where if the angle is greater than 9deg, we instead cast in from one side of the map instead of from directly overhead, but I don't even want to think about that right now.

Reply:
Ah, OK, so the offset will potentially span multiple z-levels.  That makes more sense.  In terms of in-game impact, though, I think the sideways casting would be much more relevant, because most buildings are only one or two levels high.  This version is only really going to matter for holes in roofs and things like that.

Reply:
And holes in the ground, which was the initial impetus for it.
Now you can set ofvan ANFO charge (debug immortality recommended) and when you jump into the resulting hole it'll have sunlight streaming in.